### DISCUSSION TITLE
VectorTiles: Crashes after a while when writing to file (mapbox)

### DISCUSSION THREAD

**Discussion by ludvigwiden2xp** [NetTopologySuite/NetTopologySuite#628] [Q&A]
Hi! I am implementing a tile-builder by reading geometries from .sqlite files, adding different layers to vector tiles and then writing to file. I have a TileRange with 26000 tiles and everything seems to be working fine until around 50% of the job is done (14000 VectorTiles written),  then the program crashes. I get this error.

```
Unhandled Exception: System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at NetTopologySuite.Geometries.Implementation.CoordinateArraySequence.GetOrdinate(Int32 index, Int32 ordinateIndex)
   at NetTopologySuite.IO.VectorTiles.Mapbox.TileGeometryTransform.Transform(CoordinateSequence sequence, Int32 index, Int32& currentX, Int32& currentY)
   at NetTopologySuite.IO.VectorTiles.Mapbox.MapboxTileWriter.Encode(CoordinateSequence sequence, TileGeometryTransform tgt, Int32& currentX, Int32& currentY, Boolean ring, Boolean ccw)
   at NetTopologySuite.IO.VectorTiles.Mapbox.MapboxTileWriter.<Encode>d__7.MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at NetTopologySuite.IO.VectorTiles.Mapbox.MapboxTileWriter.Write(VectorTile vectorTile, Stream stream, UInt32 extent, String idAttributeName)
```

The code pretty much look like this. In "tlb.BuildNTSLayer" I check if a geometry is invalid, if so, I use the geometryFixer.Fix(), and then perform intersection with a square geometry to remove any part (of a feature) outside of the tile.

```
foreach (var tlb in mapDataSources)
            {
                Layer layer = tlb.BuildNTSLayer(tile);
                if (!layer.IsEmpty)
                {
                    vectorTile.Layers.Add(layer);     
                }              
            }
if (!vectorTile.IsEmpty)
{
     using (var fs = new FileStream(path, FileMode.Create))
      {             
            vectorTile.Write(fs, 35000, "fid");             
       }
}
```

Do you have any ideas where this problem could be coming from?

**Comment by DGuidi** [comment]
looking at the stack, seems totally related to something happening in the [VectorTiles](https://github.com/NetTopologySuite/NetTopologySuite.IO.VectorTiles) package, and depending to some specific data.
I can only argue that the exception originates in [this line](https://github.com/NetTopologySuite/NetTopologySuite.IO.VectorTiles/blob/036e5796b81a30d6b366a78c4516169dd48737ee/src/NetTopologySuite.IO.VectorTiles.Mapbox/TileGeometryTransform.cs#L46) that is called from [here](https://github.com/NetTopologySuite/NetTopologySuite.IO.VectorTiles/blob/092c17ab8d0ccfdf0761ae4067c8cda932320270/src/NetTopologySuite.IO.VectorTiles.Mapbox/MapboxTileWriter.cs#L199)

So maybe some of your data is a `POINT EMPTY`?

**Comment by DGuidi** [comment]
I suppose that the test that reproduces the behaviour you experienced is this one:
```
[Fact]
public void TileGeometryTransform_Transform_EmptySequence()
{
	var tileGeometry = new TileGeometryTransform(new Tile(0, 0, 0), 4096);
	var x = 0;
	var y = 0;
	tileGeometry.Transform(new CoordinateArraySequence(new Coordinate[0]), 0, ref x, ref y);

	Assert.Equal(2048, x);
	Assert.Equal(2048, y);
}
```

that throws the same exception you experienced:
```
Message: 
System.IndexOutOfRangeException : Index was outside the bounds of the array.

  Stack Trace: 
CoordinateArraySequence.GetOrdinate(Int32 index, Int32 ordinateIndex)
CoordinateSequence.GetOrdinate(Int32 index, Ordinate ordinate)
TileGeometryTransform.Transform(CoordinateSequence sequence, Int32 index, Int32& currentX, Int32& currentY) line 46
TileGeometryTransformTests.TileGeometryTransform_Transform_EmptySequence() line 17
```
